File: src/engine/colorWheel.ts
import type { Color, MaterialType } from '../data/types';
import { ALL_COLORS, canMix, mixResult } from '../data/colors';

export function createEmptyWheel(): Record<Color, number> {
  const wheel: Partial<Record<Color, number>> = {};
  for (const c of ALL_COLORS) wheel[c] = 0;
  return wheel as Record<Color, number>;
}

export function createStartingMaterials(): Record<MaterialType, number> {
  return { Textiles: 0, Ceramics: 0, Paintings: 0 };
}

export function storeColor(wheel: Record<Color, number>, color: Color): void {
  wheel[color]++;
}

export function removeColor(wheel: Record<Color, number>, color: Color): boolean {
  if (wheel[color] <= 0) return false;
  wheel[color]--;
  return true;
}

export function performMix(wheel: Record<Color, number>, a: Color, b: Color): boolean {
  if (!canMix(a, b)) return false;
  if (wheel[a] <= 0 || wheel[b] <= 0) return false;
  wheel[a]--;
  wheel[b]--;
  const result = mixResult(a, b);
  wheel[result]++;
  return true;
}

export function canPayCost(wheel: Record<Color, number>, cost: Color[]): boolean {
  const temp = { ...wheel };
  for (const c of cost) {
    if (temp[c] <= 0) return false;
    temp[c]--;
  }
  return true;
}

export function payCost(wheel: Record<Color, number>, cost: Color[]): boolean {
  if (!canPayCost(wheel, cost)) return false;
  for (const c of cost) wheel[c]--;
  return true;
}

File: colori-core/src/color_wheel.rs
use crate::colors::{can_mix, mix_result};
use crate::types::{Color, ColorWheel};

#[inline]
pub fn store_color(wheel: &mut ColorWheel, color: Color) {
    wheel.increment(color);
}

#[inline]
pub fn remove_color(wheel: &mut ColorWheel, color: Color) -> bool {
    wheel.decrement(color)
}

#[inline]
pub fn perform_mix(wheel: &mut ColorWheel, a: Color, b: Color) -> bool {
    if !can_mix(a, b) {
        return false;
    }
    if wheel.get(a) == 0 || wheel.get(b) == 0 {
        return false;
    }
    wheel.decrement(a);
    wheel.decrement(b);
    let result = mix_result(a, b);
    wheel.increment(result);
    true
}

/// Performs a mix without checking `can_mix` or wheel amounts.
/// The caller must guarantee that `a` and `b` are a valid mix pair
/// and that the wheel has at least one of each.
#[inline]
pub fn perform_mix_unchecked(wheel: &mut ColorWheel, a: Color, b: Color) {
    wheel.decrement(a);
    wheel.decrement(b);
    let result = mix_result(a, b);
    wheel.increment(result);
}

#[inline]
pub fn can_pay_cost(wheel: &ColorWheel, cost: &[Color]) -> bool {
    let mut used = [0u32; 12];
    for &c in cost {
        let idx = c.index();
        let needed = used[idx] + 1;
        if wheel.get(c) < needed {
            return false;
        }
        used[idx] = needed;
    }
    true
}

#[inline]
pub fn pay_cost(wheel: &mut ColorWheel, cost: &[Color]) -> bool {
    if !can_pay_cost(wheel, cost) {
        return false;
    }
    for &c in cost {
        wheel.decrement(c);
    }
    true
}
